docs: TCP and UDP examples / documentation (#4932)
kubernetes-sigs/gateway-api: site/content/en/guides/user-guides/udp.md

---
title: "UDP routing"
weight: 14
---

Gateway API is designed to work with multiple protocols and [UDPRoute][udproute]
is one such route which allows for managing [UDP][udp] traffic.

In this example, we have one Gateway resource and two UDPRoute resources that
distribute the traffic with the following rules:

- All UDP datagrams on port 8080 of the Gateway are forwarded to port 6000 of
  `my-foo-service` Kubernetes Service.
- All UDP datagrams on port 8090 of the Gateway are forwarded to port 6000 of
  `my-bar-service` Kubernetes Service.

In this example two `UDP` listeners will be applied to the [Gateway][gateway]
in order to route them to two separate backend `UDPRoutes`, note that the
`protocol` set for the `listeners` on the `Gateway` is `UDP`:

> {{< readfile file="/examples/experimental/basic-udp.yaml" code="true" lang="yaml" >}}

Reviewer: examples/standard

---
title: "UDP routing"
weight: 14
---

Gateway API is designed to work with multiple protocols and [UDPRoute][udproute]
is one such route which allows for managing [UDP][udp] traffic.

In this example, we have one Gateway resource and two UDPRoute resources that
distribute the traffic with the following rules:

- All UDP datagrams on port 8080 of the Gateway are forwarded to port 6000 of
  `my-foo-service` Kubernetes Service.
- All UDP datagrams on port 8090 of the Gateway are forwarded to port 6000 of
  `my-bar-service` Kubernetes Service.

> 

Reviewer 2: There is no example that matches this preamble, but if all of this text refers to the embedded example, put it either above all the descriptive text, or below it,  and use "in the example below" or "in the example above".

In this example two `UDP` listeners will be applied to the [Gateway][gateway]

> in order to route them to two separate backend `UDPRoutes`, note that the

Reviewer 2: ```suggestion
in order to route traffic to two separate backend `UDPRoutes`, note that the
```

`protocol` set for the `listeners` on the `Gateway` is `UDP`:

{{< readfile file="/examples/standard/basic-udp.yaml" code="true" lang="yaml" >}}

In the above example we separate the traffic for the two separate backend UDP

> [Services][svc] by using the `sectionName` field in the `parentRefs`:

Reviewer 2: ```suggestion
[Services][svc] by using the `sectionName` field in the `parentRefs` to refer to a specific named listener:
```

```yaml
spec:
  parentRefs:
    - name: my-udp-gateway
      sectionName: foo
```

This corresponds directly with the `name` in the `listeners` in the `Gateway`:

```yaml
listeners:
  - name: foo
    protocol: UDP
    port: 8080
  - name: bar
    protocol: UDP
    port: 8090
```

In this way each `UDPRoute` "attaches" itself to a different port on the
`Gateway` so that the service `my-foo-service` is taking traffic for port `8080`

> from outside the cluster and `my-bar-service` takes the port `8090` traffic.

Reviewer 2: ```suggestion
`Gateway` so that the service `my-foo-service` receives traffic for port `8080`
and `my-bar-service` receives traffic for port `8090`.
```

Note that you can achieve this same result by binding the Routes to the Gateway
listeners using the `port` field in the `parentRefs`:

```yaml
spec:
  parentRefs:
    - name: my-udp-gateway
      port: 8080
```

Using the `port` field instead of `sectionName` for the attachment has the

> downside of more tightly coupling the relationship between the Gateway and

Reviewer 2: I think you said it better for tcpRoute.  Can you use that here?

However, when binding Routes by port number, Gateway admins will no longer have
the flexibility to switch ports on the Gateway without also updating the Routes.
This approach should only be used when a Route must bind to a specific port
number, rather than to named listeners whose ports may change.
